test_app/tests/lib/redis/test_client.py

```diff
@@ -0,0 +1,176 @@
+import copy
+from unittest import mock
+
+import pytest
+from django.core.exceptions import ImproperlyConfigured
+from django_redis.cache import RedisCache
+from redis.client import Redis
+from redis.exceptions import RedisClusterException
+
+from ansible_base.lib.redis.client import DABRedisCluster, RedisClient
+
+
+@pytest.mark.parametrize(
+    "args,clustered,clustered_hosts",
+    [
+        ({}, False, ''),
+        ({'OPTIONS': {}}, False, ''),
+        ({'OPTIONS': {'CLIENT_CLASS_KWARGS': {}}}, False, ''),
+        ({'OPTIONS': {'CLIENT_CLASS_KWARGS': {'clustered': False}}}, False, ''),
+        ({'OPTIONS': {'CLIENT_CLASS_KWARGS': {'clustered': True}}}, True, ''),
+        ({'OPTIONS': {'CLIENT_CLASS_KWARGS': {'clustered_hosts': ''}}}, False, ''),
+        ({'OPTIONS': {'CLIENT_CLASS_KWARGS': {'clustered_hosts': 'a:1'}}}, False, 'a:1'),
+    ],
+)
+def test_redis_client_init(args, clustered, clustered_hosts):
+    redis_cache = RedisCache('localhost', args)
+    client = RedisClient('localhost', args, redis_cache)
+    assert client.clustered is clustered
+    assert client.clustered_hosts == clustered_hosts
+
+
+def test_redis_client_confirm_connect_does_not_change_options():
+    args = {'OPTIONS': {'CLIENT_CLASS_KWARGS': {'clustered': False, 'clustered_hosts': 'a:1'}}}
+    validate_args = copy.deepcopy(args)
+    redis_cache = RedisCache('localhost', args)
+    client = RedisClient('localhost', args, redis_cache)
+    client.connect()
+    assert args == validate_args
+
+
+@pytest.mark.parametrize(
+    "url,host,port,username,password,db",
+    [
+        ('localhost', 'localhost', 6379, None, None, 0),
+        ('redis://my_mom:her_pass@home.com:1234/1', 'home.com', 1234, 'my_mom', 'her_pass', 1),
+        ('redis://my_mom:her_pass@home.com/1', 'home.com', 6379, 'my_mom', 'her_pass', 1),
+        ('redis://my_mom@home.com/', 'home.com', 6379, 'my_mom', None, 0),
+        ('redis://:her_pass@home.com/', 'home.com', 6379, None, 'her_pass', 0),
+        ('redis://home.com:1234/a', 'home.com', 1234, None, None, 0),
```

Comment:
Nitpick, but I try to avoid using "real-ish" URLs in tests, just to avoid any future concerns. I tend to use `example.com` (which is IANA-reserved for this kind of stuff).

Reply:
I'll change this to `example.com`, I should know better :)